File: pickup-server/src/main/java/me/synology/hajubal/coins/crawler/impl/ruliweb/RuliwebCrawler.java
package me.synology.hajubal.coins.crawler.impl.ruliweb;

import java.io.IOException;
import java.util.Set;
import me.synology.hajubal.coins.config.CrawlerProperties;
import me.synology.hajubal.coins.crawler.AbstractWebCrawler;
import me.synology.hajubal.coins.crawler.PointPostUrlFetcher;
import me.synology.hajubal.coins.crawler.SiteData;
import me.synology.hajubal.coins.respository.SiteRepository;
import org.springframework.stereotype.Component;

/**
 * 루리웹 사이트 핫딜 게시판 포인트 URL 크롤러
 */
@Component
public class RuliwebCrawler extends AbstractWebCrawler {

  private final PointPostUrlFetcher pointPostUrlFetcher;
  private final RuliwebSiteData ruliwebSiteData;

  public RuliwebCrawler(
      PointPostUrlFetcher pointPostUrlFetcher,
      SiteRepository siteRepository,
      CrawlerProperties crawlerProperties,
      RuliwebSiteData ruliwebSiteData) {
    super(siteRepository, crawlerProperties);
    this.pointPostUrlFetcher = pointPostUrlFetcher;
    this.ruliwebSiteData = ruliwebSiteData;
  }

  @Override
  protected SiteData getSiteData() {
    return ruliwebSiteData;
  }

  @Override
  protected String getArticleSelector() {
    return ".board_main_view a";
  }

  @Override
  protected Set<String> fetchPostUrls(String siteUrl) throws IOException {
    return pointPostUrlFetcher.fetchPostUrls(siteUrl);
  }
}

File: src/modules/crawler/crawlers/ruliweb.crawler.ts
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import * as cheerio from 'cheerio';
import { BaseCrawler, SiteData } from './base.crawler';

@Injectable()
export class RuliwebCrawler extends BaseCrawler {
  constructor(configService: ConfigService) {
    super(configService, RuliwebCrawler.name);
  }

  getSiteData(): SiteData {
    return {
      siteName: '루리웹',
      domain: 'https://m.ruliweb.com',
      boardUrl: '/ps/board/1020',
    };
  }

  getArticleSelector(): string {
    return '.board_main_view a';
  }

  async fetchPostUrls(siteUrl: string): Promise<Set<string>> {
    const postUrls = new Set<string>();

    try {
      const html = await this.fetchWithRetry(siteUrl);
      const $ = cheerio.load(html);

      // Find all post links on Ruliweb mobile
      $('a.subject_link, a.title_text').each((_, element) => {
        const href = $(element).attr('href');
        if (href && (href.includes('/ps/board/') || href.includes('/bbs/board/'))) {
          postUrls.add(href);
        }
      });

      // Also check for desktop view links
      $('a.deco').each((_, element) => {
        const href = $(element).attr('href');
        if (href && (href.includes('/ps/board/') || href.includes('/bbs/board/'))) {
          postUrls.add(href);
        }
      });

      this.logger.debug(`Found ${postUrls.size} post URLs from Ruliweb`);
    } catch (error) {
      this.logger.error(`Failed to fetch post URLs from Ruliweb`, error);
    }

    return postUrls;
  }
}
